extras/devices/LVFS/LVFS.py

```diff
@@ -0,0 +1,332 @@
+import subprocess
+import json
+import time
+from devices.base import AbstractDevice, logger
+
+SSH_OPTS = "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+
+
+class LVFSDevice(AbstractDevice):
+    fw_update_type = "LVFS"
+    vendor = ["HP", "Dell Inc.", "LENOVO"]
+    reboot_timeout = 600
+
+    def __init__(self, ipaddr, user, password):
+        super().__init__(ipaddr, user, password)
+        # self.run_cmd("sudo apt update && sudo apt install fwupd")
+
+    def run_cmd(self, cmd, raise_stderr=True, timeout=30):
+        """
+        Execute command on DUT via SSH
+
+        :returns: return code, stdout, stderr
+        :rtype: int, string, string
+        """
+        if self.password == "":
+            ssh_cmd = 'ssh -t %s %s@%s "%s"' % (
+                SSH_OPTS,
+                self.user,
+                self.ipaddr,
+                cmd,
+            )
+        else:
+            ssh_cmd = (
+                'sshpass -p %s  ssh -t %s %s@%s "%s"'
+                % (self.password, SSH_OPTS, self.user, self.ipaddr, cmd),
+            )
+        r = subprocess.run(
+            ssh_cmd,
+            shell=True,
+            capture_output=True,
+            timeout=timeout,
+        )
+        rc, stdout, stderr = (
+            r.returncode,
+            r.stdout.decode().strip(),
+            r.stderr.decode().strip(),
+        )
+        logger.debug("Run command: %s" % ssh_cmd)
+        if raise_stderr and rc != 0:
+            err_msg = "Failed to execute %s:\n [%d] %s %s" % (
+                cmd,
+                rc,
+                stdout,
+                stderr,
+            )
+            logger.debug(err_msg)
+            raise RuntimeError(err_msg)
+        else:
+            return rc, stdout, stderr
+
+    def get_fw_info(self):
+        """
+        Get current firmware version of all updatable devices on DUT
+        And print out devices with upgradable/downgradable versions
+        """
+        logger.info("collect firmware info")
+        self.run_cmd("sudo fwupdmgr refresh --force")
+        rc, stdout, stderr = self.run_cmd("sudo fwupdmgr get-devices --json")
+        logger.debug("$fwupdmgr get-devices = \n%s" % stdout)
+        for dev in json.loads(stdout)["Devices"]:
+            if "Flags" in dev:
+                if "updatable" in dev["Flags"]:
```

Comment:
You can unite these two conditions into one with “and”

Reply:
Fixed.